src/Bicep.Core/Syntax/OutputDeclarationSyntax.cs

```diff
@@ -7,30 +7,35 @@ namespace Bicep.Core.Syntax
 {
     public class OutputDeclarationSyntax : SyntaxBase, IDeclarationSyntax
     {
-        public OutputDeclarationSyntax(Token outputKeyword, IdentifierSyntax name, TypeSyntax type, Token assignment, SyntaxBase value)
+        public OutputDeclarationSyntax(Token keyword, IdentifierSyntaxBase name, SyntaxBase type, SyntaxBase assignment, SyntaxBase value)
         {
-            AssertKeyword(outputKeyword, nameof(outputKeyword), LanguageConstants.OutputKeyword);
-            AssertTokenType(assignment, nameof(assignment), TokenType.Assignment);
+            AssertKeyword(keyword, nameof(keyword), LanguageConstants.OutputKeyword);
+            AssertSyntaxType(name, nameof(name), typeof(IdentifierSyntax), typeof(MalformedIdentifierSyntax));
+            AssertSyntaxType(type, nameof(type), typeof(TypeSyntax), typeof(SkippedTriviaSyntax));
+            AssertSyntaxType(assignment, nameof(assignment), typeof(Token), typeof(SkippedTriviaSyntax));
+            AssertTokenType(assignment as Token, nameof(assignment), TokenType.Assignment);
 
-            this.OutputKeyword = outputKeyword;
+            this.Keyword = keyword;
             this.Name = name;
             this.Type = type;
             this.Assignment = assignment;
             this.Value = value;
         }
 
-        public Token OutputKeyword { get; }
+        public Token Keyword { get; }
 
-        public IdentifierSyntax Name { get; }
+        public IdentifierSyntaxBase Name { get; }
 
-        public TypeSyntax Type { get; }
+        public SyntaxBase Type { get; }
 
-        public Token Assignment { get; }
+        public SyntaxBase Assignment { get; }
 
         public SyntaxBase Value { get; }
 
         public override void Accept(SyntaxVisitor visitor) => visitor.VisitOutputDeclarationSyntax(this);
 
-        public override TextSpan Span => TextSpan.Between(OutputKeyword, Value);
+        public override TextSpan Span => TextSpan.Between(Keyword, Value);
+
+        public TypeSyntax? OutputType => this.Type as TypeSyntax;
```

Comment:
As we're introducing more and more of these cases where we need to check if something is valid and then cast it, can you think of any way to commonize this pattern and improve on type safety?

Reply:
I'll have to think about it... No changes for now.